x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,180 @@
-	// if !remainingGammShares.IsZero() {
-	// 	newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
-	// 	newLockId = newLock.ID
-	// 	if err != nil {
-	// 		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// 	}
-	// 	// If the lock was unlocking, we begin the unlock from where it left off.
-	// 	if preUnlockLock.IsUnlocking() {
-	// 		_, err := k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
-	// 		if err != nil {
-	// 			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// 		}
-	// 	}
-	// }
-
-	return positionId, amount0, amount1, liquidity, joinTime, poolIdLeaving, poolIdEntering, newLockId, nil
+func (k Keeper) MigrateLockedPositionFromBalancerToConcentrated(ctx sdk.Context, sender sdk.AccAddress, lockId uint64, sharesToMigrate sdk.Coin) (positionId uint64, amount0, amount1 sdk.Int, liquidity sdk.Dec, joinTime time.Time, poolIdLeaving, poolIdEntering, gammLockId, concentratedLockId uint64, err error) {
+	// Get the balancer poolId by parsing the gamm share denom.
+	poolIdLeaving = gammtypes.MustGetPoolIdFromShareDenom(sharesToMigrate.Denom)
+
+	// Ensure a governance sanctioned link exists between the balancer pool and the concentrated pool.
+	poolIdEntering, err = k.gk.GetLinkedConcentratedPoolID(ctx, poolIdLeaving)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// Get the concentrated pool from the provided ID and type cast it to ConcentratedPoolExtension.
+	concentratedPool, err := k.clk.GetPoolFromPoolIdAndConvertToConcentrated(ctx, poolIdEntering)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// Check that lockID corresponds to sender, and contains correct denomination of LP shares.
+	lock, err := k.validateLockForUnpool(ctx, sender, poolIdLeaving, lockId)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+	gammSharesInLock := lock.Coins[0]
+	preUnlockLock := *lock
+
+	// Before we break the lock, we must note the time remaining on the lock.
+	remainingLockTime := k.getExistingLockRemainingDuration(ctx, lock)
+
+	// Check if the lock has a corresponding synthetic lock.
+	// Synthetic lock existence implies that the lock is superfluid delegated or undelegating.
+	synthLockBeforeMigration := k.lk.GetAllSyntheticLockupsByLockup(ctx, lockId)
+
+	// If it does, check if it is superfluid delegated or undelegating.
+	wasSuperfluidUndelegatingBeforeMigration := len(synthLockBeforeMigration) > 0 && strings.Contains(synthLockBeforeMigration[0].SynthDenom, "superunbonding")
+	wasSuperfluidDelegatedBeforeMigration := len(synthLockBeforeMigration) > 0 && strings.Contains(synthLockBeforeMigration[0].SynthDenom, "superbonding")
+	if wasSuperfluidUndelegatingBeforeMigration && wasSuperfluidDelegatedBeforeMigration {
+		// This is a defense in depth measure and should never happen.
```

Comment:
Can we also add defense in depth measure for both not being true?

Reply:
Both not being true is fine, as this implies we are migrating a non superfluid lock. This is more clear now since I refactored into three distinct methods.